src/main/scala/mesosphere/marathon/core/volume/providers/ResidentVolumeProvider.scala

```diff
@@ -0,0 +1,40 @@
+package mesosphere.marathon.core.volume.providers
+
+import com.wix.accord.Validator
+import com.wix.accord.combinators.NilValidator
+import com.wix.accord.dsl._
+import mesosphere.marathon.core.volume._
+import mesosphere.marathon.state._
+
+/**
+  * ResidentVolumeProvider handles persistent volumes allocated from agent resources.
+  */
+protected[volume] case object ResidentVolumeProvider
+    extends AbstractPersistentVolumeProvider("resident") {
+
+  import mesosphere.marathon.api.v2.Validation._
+  import org.apache.mesos.Protos.Volume.Mode
+
+  // no provider-specific rules at the app level
+  // TODO(jdef) could/should refactor resident validation from AppDefinition to here
+  val appValidation: Validator[AppDefinition] = new NilValidator[AppDefinition]
+
+  // no provider-specific rules at the group level
+  val groupValidation: Validator[Group] = new NilValidator[Group]
+
+  val volumeValidation = validator[PersistentVolume] { v =>
+    v.persistent.size is notEmpty
+    v.mode is equalTo(Mode.RW)
+    //persistent volumes require those CLI parameters provided
+    v is configValueSet("mesos_authentication_principal", "mesos_role", "mesos_authentication_secret_file")
```

Comment:
it used to be in state/Volume. I don't like it in either place. do you have a suggestion for relocating this?